### ISSUE TITLE
Attention mask does not make sense

### ISSUE THREAD

**Issue by Boltzmachine** [SpatialVLA/SpatialVLA#59]
Hi, I found the `_update_causal_mask` method of the `SpatialVLAForConditionalGeneration` does not make sense to me

https://github.com/SpatialVLA/SpatialVLA/blob/18fd74b2a633ec8d9ec7aadcd803969555cc9fbd/model/modeling_spatialvla.py#L294

If you check the very first iteration when inference (when there is no kv cache yet), the zeroing operation will make the first `sequence_length` entries in each row be zeroed out, which means each query before `sequence_length` is allowed to attend tokens after it.

I check the attention in the first forward pass in inference, the causal mask is also not causal.

**Comment by DelinQu** [comment]
Hi,
In common [transformer implementations](https://github.com/huggingface/transformers/blob/1580f6465347df4e8d7c12cd4d3dc603b8a689ad/src/transformers/integrations/eager_paged.py#L19), attention weights are modified by:
```python
attn_weights = attn_weights + attention_mask
```
This means a value of 0 in the mask allows attention (no modification to the raw weights), and large negative value (like -inf) blocks attention (pushing the softmax probability to 0). The line causal_mask[:, :sequence_length] = 0.0 actually enables attention to these positions rather than blocking them.

However, if Flash Attention is used, the behavior changes: 
The attention_mask is only used for padding handling, not causal masking. Causal attention is controlled separately via the is_causal flag, which is implemented efficiently in CUDA.

See the relevant HuggingFace implementation [here](https://github.com/huggingface/transformers/blob/5d7739f15a6e50de416977fe2cc9cb516d67edda/src/transformers/modeling_flash_attention_utils.py#L310).

<img width="1061" height="508" alt="Image" src="https://github.com/user-attachments/assets/46139f7c-92dc-471a-9d8b-d4c763213579" />

**Comment by Boltzmachine** [comment]
> The line causal_mask[:, :sequence_length] = 0.0 actually enables attention to these positions rather than blocking them.

Yeah that is exactly what I said what is wrong. Let's say the causal_mask is of size `(300, 500)` where `300` is the sequence length and `500` is the `target_length`, which is obtained by `target_length = past_key_values.get_max_cache_shape()` in previous lines because gemma use a `HybridCache`.

In this case, the operation `causal_mask[:, :300] = 0` will make the token attend to tokens after it. For example, Now both `causal_mask[0, 299]` and `causal_mask[299, 0]` are zero, which means the 299th token and the 0th token are visible to each other, making a bidirectional attention.

**Comment by Boltzmachine** [comment]
Hi just a follow up @DelinQu. I really appreciate it if you could help with this.
